sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -35,15 +35,14 @@ case class DataSourceV2Relation(
     options: Map[String, String],
     projection: Seq[AttributeReference],
     filters: Option[Seq[Expression]] = None,
-    userSpecifiedSchema: Option[StructType] = None) extends LeafNode with MultiInstanceRelation {
+    userSpecifiedSchema: Option[StructType] = None)
+  extends LeafNode with MultiInstanceRelation with DataSourceV2QueryPlan {
```

Comment:
Extending `DataSourceV2QueryPlan` changes the definition of equality, which I don't think is correct. At a minimum, I would say that two relations are equal if they produce the same sequence of records. Equality as implemented in this PR would allow completely different folders of data to be considered equal as long as they produce the same schema and have the same filters.

In general, I don't see the utility of `DataSourceV2QueryPlan` for the purpose of this PR, which is to improve explain results. This doesn't use the source name if it is named, which is a regression in the explain results. It also doesn't indicate that the source is v2. And finally, it doesn't show the most important part of the scan, which is where the data comes from.

Reply:
Sorry I didn't realize you removed the extending of `DataSourceReaderHolder`, otherwise I would point it out in your PR. In general the question is, how should we define the equality of data source relation? It was defined by `output, reader.getClass, filters` before, and your PR changed it to the default equality of `DataSourceV2Relation` silently.

The major difference is, should `options` take part in the equality? The answer is obviously yes, like `path`, so I'll add `options` to the equality.

BTW `DataSourceV2QueryPlan` is needed, as there are 3 plans need to implement explain: `DataSourceV2Relation`, `StreamingDataSourceV2Relation`, `DataSourceV2ScanExec`.

> This doesn't use the source name if it is named

I missed that, let me add back.

> doesn't indicate that the source is v2

I'll improve it

> it doesn't show the most important part of the scan

Do you mean the path option? First I don't think showing all the options is a good idea, as it can be a lot. My future plan is to show these standard options like path, table, etc. Again this is something added silently, there is no consensus about how to explain a data source v2 relation, my PR tries to have people focus on this part and have a consensus.

Reply:
Part of the utility of an immutable plan is to make equality work correctly. I thought that was clear, but sorry if it was not. Maybe I should have pointed it out explicitly.

Equality should definitely be based on all of the inputs that affect the output rows. That includes location or table identifier, user schema, filters, requested projection, implementation, and other options. I think case class equality is correct here.

> there is no consensus about how to explain a data source v2 relation, my PR tries to have people focus on this part and have a consensus.

That's a good goal for this PR, so I think you should roll back the changes to the relation so equality is not affected by it.

> Do you mean the path option?

Yes. The data that will be scanned is an important part. As you know, I think this should be part of the relation itself and not options. That would solve this problem.